src/dos/cdrom_image.cpp

```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
+{
+	const Bit16u bytes = Sound_Decode(sample);
+	memcpy(buffer, sample->buffer, bytes);
+	return bytes;
+}
+
+Bit16u CDROM_Interface_Image::AudioFile::getEndian()
+{
+	return sample->actual.format;
+}
+
+Bit32u CDROM_Interface_Image::AudioFile::getRate()
+{
+	Bit32u rate(0);
+	if (sample) {
+		rate = sample->actual.rate;
 	}
-	lastSeek = seek;
-	int bytes = Sound_Decode(sample);
-	if (bytes < count) {
-		memcpy(buffer, sample->buffer, bytes);
-		memset(buffer + bytes, 0, count - bytes);
-	} else {
-		memcpy(buffer, sample->buffer, count);
+	return rate;
+}
+
+Bit8u CDROM_Interface_Image::AudioFile::getChannels()
+{
+	Bit8u channels(0);
+	if (sample) {
+		channels = sample->actual.channels;
 	}
-	
-	return !(sample->flags & SOUND_SAMPLEFLAG_ERROR);
+	return channels;
 }
 
 int CDROM_Interface_Image::AudioFile::getLength()
 {
-	int time = 1;
-	int shift = 0;
-	if (!(sample->flags & SOUND_SAMPLEFLAG_CANSEEK)) return -1;
-	
-	while (true) {
-		int success = Sound_Seek(sample, (unsigned int)(shift + time));
-		if (!success) {
-			if (time == 1) return lround((double)shift * 176.4f);
-			shift += time >> 1;
-			time = 1;
-		} else {
-			if (time > ((numeric_limits<int>::max() - shift) / 2)) return -1;
-			time = time << 1;
-		}
+	int length(-1);
+
+	// GetDuration returns milliseconds ... but getLength needs Red Book bytes.
+	const int duration_ms = Sound_GetDuration(sample);
+	if (duration_ms > 0) {
+		// ... so convert ms to "Red Book bytes" by multiplying with 176.4f,
+		// which is 44,100 samples/second * 2-channels * 2 bytes/sample
+		// / 1000 milliseconds/second
+		length = (int) round(duration_ms * 176.4f);
 	}
-}
+#ifdef DEBUG
+	LOG_MSG("%s CDROM: AudioFile::getLength is %d bytes", get_time(), length);
 #endif
 
+	return length;
+}
+
 // initialize static members
 int CDROM_Interface_Image::refCount = 0;
 CDROM_Interface_Image* CDROM_Interface_Image::images[26] = {};
 CDROM_Interface_Image::imagePlayer CDROM_Interface_Image::player = {
-	NULL, NULL, NULL, {0}, 0, 0, 0, false, false, false, {0} };
-
+	NULL,  // CDROM_Interface_Image*
+	NULL,  // MixerChannel*
+	{0},   // buffer[]
+	0,     // startFrame
+	0,     // currFrame
+	0,     // numFrames
+	false, // isPlaying
+	false, // isPaused
+	false, // ctrlUsed
+	{0},   // ctrlData struct
+	NULL,  // activeTrack
+	NULL,  // addSamples
+	0,     // playbackTotal
+	0,     // playbackRemaining
+	0,     // bufferPos
+	0      // bufferConsumed
+};
 	
 CDROM_Interface_Image::CDROM_Interface_Image(Bit8u subUnit)
-                      :subUnit(subUnit)
+		      :subUnit(subUnit)
 {
 	images[subUnit] = this;
 	if (refCount == 0) {
-		player.mutex = SDL_CreateMutex();
-		if (!player.channel) {
-			player.channel = MIXER_AddChannel(&CDAudioCallBack, 44100, "CDAUDIO");
+		if (player.channel == NULL) {
+			// channel is kept dormant except during cdrom playback periods
+			player.channel = MIXER_AddChannel(&CDAudioCallBack, 0, "CDAUDIO");
+			player.channel->Enable(false);
```

Comment:
*Idea fo future improvement*: Method `Enable(bool)` is weird, it would be nicer to have methods `Enable()`, `Disable()`.

Reply:
Absolutely. A series of dedicated mass-cleanup patches are probably in order; where each commit or PR changes one specific aspect or style, across the board.

Throw in C++11'isms (say, changing all news & mallocs to unique-pointers, or arrays to vectors, and so on), would be fantastic and likely produce large reductions in LOC and save on complexity.  (side-comment: would be great if we had a cyclomatic complexity analysis check along side the static analysis. Will have to dig around and see what's out there).